### PR TITLE
feat: add virtiofs module check to image verify for OCI VM images

### PR DESCRIPTION
**Pull request by CMGS**
## Summary

- Add virtiofs kernel module detection to `cocoon image verify` for OCI VM images
- When verifying an OCI VM image, the kernel layer tar is inspected for an initramfs entry, which is then scanned (via cpio newc parsing) for `virtiofs.ko` (including `.xz`, `.zst`, `.gz` variants)
- Results are reported as `VirtiofsChecked`/`VirtiofsFound` fields on `BootCheckResult` and surfaced as PASS/WARN messages in the verify output
- Implements the Section 6.5 requirement from docs/04.1 (initramfs virtiofs support check)

## Changes

| File | Description |
|------|-------------|
| `oci/initramfs.go` | New: `CheckInitramfsVirtiofs()` — opens kernel layer tar, finds initrd entry, decompresses (gzip/uncompressed), scans cpio newc for virtiofs.ko |
| `oci/initramfs_test.go` | Tests: found/not-found/uncompressed/zstd-error/missing-initrd cases plus unit tests for cpio helpers |
| `cmd/cocoon/images.go` | Wire `checkInitramfsVirtiofs()` into `verifyOCILayoutBootability()` |
| `cmd/cocoon/images_test.go` | Integration tests: no-kernel-layer, nil-info, bad-blob-path edge cases |
| `image/types.go` | Add `VirtiofsFound`/`VirtiofsChecked` fields to `BootCheckResult` |

## Design decisions

- **Minimal cpio parser**: Hand-rolled newc format parser instead of importing a full cpio library, keeping dependencies lean
- **Gzip decompression only**: Gzip is supported natively; zstd returns an informative error (avoiding a cgo or large pure-Go dependency)
- **Non-blocking**: Virtiofs absence is a WARN, not an error — the image is still considered bootable since virtiofs may be compiled into the kernel or loaded via other means

## Test plan

- [x] `make fmt` — passes
- [x] `make lint` — 0 issues (both linux and darwin)
- [x] `make test` — all tests pass with race detector
- [ ] Manual: `cocoon image verify <oci-tag>` on an image with virtiofs in initramfs → should show "virtiofs module found in initramfs"
- [ ] Manual: `cocoon image verify <oci-tag>` on an image without virtiofs → should show warning about missing virtiofs

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by CMGS — COMMENTED** [review]
## Review: feat: add virtiofs module check to image verify for OCI VM images

### Verdict: Approve with suggestions

Overall this is a clean, well-structured addition. The cpio newc parser is correct per the spec, error handling degrades gracefully, and test coverage is strong across the major paths. CI lint passes; test job is pending but the suite passed locally per the PR description.

### Verified

- **cpio newc parser correctness**: Header layout matches `man 5 cpio` -- magic(6) + 13x8-byte hex fields = 110 bytes. Field offsets for `namesize` (offset 94) and `filesize` (offset 54) are correct. Alignment padding uses the standard `(n+3) &^ 3` idiom. Both `070701` (newc) and `070702` (CRC) magic are accepted. Trailer sentinel `TRAILER!!!` terminates scanning. This is correct.
- **Path traversal safety**: The tar reader (`archive/tar`) is used only to locate a known entry name; no extraction to disk occurs. The cpio parser only reads filenames for string matching -- no path-based file operations. `LayoutBlobPath` validates digest format via `ParseSHA256Digest` (rejects non-hex), preventing path traversal through the digest. No security concerns here.
- **Graceful degradation**: Errors from `CheckInitramfsVirtiofs` are surfaced as warnings, not hard failures. `VirtiofsChecked=false` is set on error. Missing kernel layer silently skips. Missing initrd returns an error that the caller downgrades to a warning. zstd returns an informative error. Nil-info is handled. This is well done.
- **Edge cases**: zstd detection via magic bytes with informative error. Uncompressed cpio fallback. Multiple initrd entry names for forward compat. Compressed module variants (.xz, .zst, .gz) detected via suffix matching. Corrupt/truncated cpio data causes the parser to return false (not panic).
- **Test quality**: 8 test functions in `oci/initramfs_test.go` covering found/not-found/uncompressed/zstd-error/missing-initrd plus unit tests for helpers. 3 integration tests in `cmd/cocoon/images_test.go` for no-kernel-layer, nil-info, bad-blob-path. The test cpio builder (`buildNewcCPIO`) is a faithful newc writer, which validates the parser against real-format data. All tests use `t.Parallel()`.

### Observations (non-blocking)

1. **Unbounded `io.ReadAll` on initramfs**: Line 52 reads the entire initramfs into memory. Typical initramfs images are 30-80 MB; unusually large ones could reach 200+ MB. This is acceptable for a CLI verify tool running on the host, but worth noting for future awareness. A follow-up could add an `io.LimitReader` guard (e.g., 512 MB cap) to fail fast on corrupt tar entries with inflated sizes.

2. **Positive result uses Warnings slice**: "virtiofs module found in initramfs" is appended to `result.Warnings` (line 975). Semantically this is a PASS, not a warning. Consider using a separate info/messages field in a future iteration, or at minimum using a "PASS:" prefix to distinguish it from actual warnings in the output.

3. **`containsVirtiofsModule` allocates on every call**: The `suffixes` slice is allocated inside the function body. Since this is called per cpio entry, moving it to a package-level var would be a minor efficiency improvement. Non-blocking given the context (CLI tool, bounded entry count).

4. **CI status**: lint passes, test job is still pending at review time. Approval is conditional on the test job going green.

File: oci/initramfs.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+		}
+		if readErr != nil {
+			return false, fmt.Errorf("read kernel layer tar: %w", readErr)
+		}
+		if !isInitrdEntry(hdr.Name) {
+			continue
+		}
+
+		// Found the initramfs entry. Read it fully so we can inspect it.
+		initrdData, readAllErr := io.ReadAll(tr)
```
Comment (CMGS) [inline oci/initramfs.go]: suggestion: `io.ReadAll(tr)` reads the entire initramfs into memory with no size cap. Typical initramfs images are 30-80 MB, but a corrupt tar entry could claim an enormous filesize. Consider wrapping with `io.LimitReader` to fail fast:

```go
const maxInitramfsSize = 512 << 20 // 512 MB
initrdData, readAllErr := io.ReadAll(io.LimitReader(tr, maxInitramfsSize))
```

Non-blocking for a host-side CLI tool, but a good defensive measure.

File: oci/initramfs.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+	return false
+}
+
+// containsVirtiofsModule returns true if the path contains a virtiofs kernel
+// module filename (virtiofs.ko, virtiofs.ko.xz, virtiofs.ko.zst, virtiofs.ko.gz).
+func containsVirtiofsModule(path string) bool {
+	// Match by checking if any path segment ends with a virtiofs module name.
+	suffixes := []string{
```
Comment (CMGS) [inline oci/initramfs.go]: nit: The `suffixes` slice is allocated on every call to `containsVirtiofsModule`. Since this runs once per cpio entry, consider hoisting it to a package-level `var` to avoid the per-call allocation:

```go
var virtiofsModuleSuffixes = []string{
	"virtiofs.ko",
	"virtiofs.ko.xz",
	"virtiofs.ko.zst",
	"virtiofs.ko.gz",
}
```

File: cmd/cocoon/images.go [hunk 3]
```diff
@@ -926,7 +926,57 @@ func verifyOCILayoutBootability(layoutPath string) (*image.BootCheckResult, erro
...
+		result.VirtiofsChecked = false
+		result.Warnings = append(result.Warnings,
+			fmt.Sprintf("virtiofs check inconclusive: %v", checkErr))
+		return
+	}
+
+	result.VirtiofsChecked = true
+	result.VirtiofsFound = found
+	if found {
+		result.Warnings = append(result.Warnings, "virtiofs module found in initramfs")
```
Comment (CMGS) [inline cmd/cocoon/images.go]: suggestion: "virtiofs module found in initramfs" is a positive result, but it is appended to `result.Warnings`. This makes the verify output semantically confusing -- the user sees a "Warning" that is actually good news. Two options:

1. Prefix with `PASS:` (e.g., `"PASS: virtiofs module found in initramfs"`) so it is distinguishable in the output.
2. In a future iteration, add an `Info []string` field to `BootCheckResult` for informational messages.

Non-blocking -- current behavior is functional.

File: oci/initramfs.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	return false, fmt.Errorf("initramfs entry not found in kernel layer tar (tried: %s)",
+		strings.Join(initrdEntryNames, ", "))
+}
+
+// isInitrdEntry checks whether a tar header name matches one of the known
+// initramfs filenames (with or without a leading "./" or "/" prefix).
+func isInitrdEntry(name string) bool {
+	// Normalize: strip leading "./" or "/" to compare bare names.
+	clean := strings.TrimPrefix(strings.TrimPrefix(name, "./"), "/")
```
Comment (CMGS) [inline oci/initramfs.go]: question: `isInitrdEntry` strips `./` then `/` via chained `TrimPrefix`. This correctly handles `./initrd.img` and `/initrd.img`, but it would not match `././initrd.img` (double dot-slash). In practice this never happens in tar archives, so this is fine. Just confirming: the test at `initramfs_test.go:272` (`subdir-initrd` -> `boot/initrd.img` -> false) documents that subdirectory paths are intentionally excluded. Is that the correct expectation for all kernel layer tar formats, or could some builders nest the initrd under `boot/`?

File: oci/initramfs.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		return false, fmt.Errorf("decompress initramfs: %w", err)
+	}
+	return cpioContainsVirtiofs(raw), nil
+}
+
+// decompressInitramfs detects the compression format by magic bytes and
+// returns the decompressed payload. Supported: gzip, zstd (magic 0x28B52FFD),
+// and uncompressed (assumed cpio).
+func decompressInitramfs(data []byte) ([]byte, error) {
+	if len(data) < 4 {
```
Comment (CMGS) [inline oci/initramfs.go]: nit: The `len(data) < 4` check is good for preventing out-of-bounds on magic byte detection. However, if the data is exactly 4-6 bytes but not gzip/zstd, it falls through to the uncompressed cpio path, where `cpioContainsVirtiofs` will immediately fail the `offset+headerLen <= len(data)` check (since headerLen=110) and return false. This is correct behavior, but the error path is silent -- the caller gets `(false, nil)` rather than an error. This is acceptable since it means "no virtiofs found" which is the correct interpretation of garbage data.

File: oci/initramfs_test.go [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+		t.Fatal("expected virtiofs.ko.xz to be found in uncompressed cpio")
+	}
+}
+
+func TestCheckInitramfsVirtiofs_ZstdReturnsError(t *testing.T) {
+	t.Parallel()
+
+	// Simulate a zstd-compressed initramfs (magic bytes 0x28B52FFD + dummy data).
+	zstdMagic := []byte{0x28, 0xB5, 0x2F, 0xFD}
+	initrdData := append(zstdMagic, make([]byte, 100)...)
```
Comment (CMGS) [inline oci/initramfs_test.go]: nit: `append(zstdMagic, make([]byte, 100)...)` modifies the underlying array of `zstdMagic` if its capacity allows. Since `zstdMagic` is a literal `[]byte{0x28, 0xB5, 0x2F, 0xFD}` with len=cap=4, this is safe here. But for clarity and robustness, consider using `slices.Concat`:

```go
initrdData := slices.Concat(zstdMagic, make([]byte, 100))
```

or pre-allocating:
```go
initrdData := make([]byte, 0, 104)
initrdData = append(initrdData, zstdMagic...)
initrdData = append(initrdData, make([]byte, 100)...)
```
